src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -482,6 +492,55 @@ private MetaData loadState() throws Exception {
         return metaDataBuilder.build();
     }
 
+    /**
+     * This method tries to open all of the nodes shards and read it's segment info to ensure the lucene version
+     * of this node can read the shard before we even start it up. If the index format is too new / too old this method
+     * fails and fails the node startup.
+     */
+    private void ensureAllIndicesAreSupported(MetaData state) throws IOException {
+        if (state.indices().isEmpty()) {
+            /**
+             * This is a fallback for nodes that used to be non-master nodes. Previously
+             * nodes didn't write the index / global state. In that case we fall back to
+             * walking the data directories to find the index shards and check them.
+             */
+            final Settings settings = ImmutableSettings.builder().build(); // empty on purpose but bypassing the assertion
+            for (String index : nodeEnv.findAllIndices()) {
+                Set<ShardId> allShardIds = nodeEnv.findAllShardIds(new Index(index));
+                openShards(settings, allShardIds);
+            }
+        } else {
+            for (IndexMetaData indexMetaData : state) {
+                Set<ShardId> allShardIds = nodeEnv.findAllShardIds(new Index(indexMetaData.index()));
+                openShards(indexMetaData.getSettings(), allShardIds);
+            }
+        }
+    }
+
+    private void openShards(Settings settings, Set<ShardId> allShardIds) throws IOException {
+        for (ShardId shard : allShardIds) {
+            Path[] paths = nodeEnv.shardDataPaths(shard, settings);
+            List<Directory> dirs = new ArrayList<>();
+            try {
+                for (Path path : paths) {
+                    dirs.add(FSDirectory.open(path.resolve("index")));
+                }
+                try (DistributorDirectory dir = new DistributorDirectory(dirs.toArray(new Directory[0]))) {
+                    Lucene.readSegmentInfos(dir); // reading the seg info is enough to throw the relevant exceptions
+                } catch (IndexFormatTooOldException ex) {
+                    logger.error("Failed to open index [{}] the lucene version is too old and is not supported anymore please upgrade the index first", ex, shard.index());
+                    throw ex;
+                } catch (IndexFormatTooNewException ex) {
+                    logger.error("Failed to open index [{}] the lucene version is too new downgrading is not supported", ex, shard.index());
+                    throw ex;
+                }
```

Comment:
This is the way the check should work, but I'm afraid its not totally complete.

Imagine a 4.x commit point with some 3.x segments. I actually think lucene 5 does the wrong thing here right now and will throw some IllegalArgumentException or SPI error. I will look into it. In the future, this would be a cool test index for us to have cc @rjernst